src/conversations/Conversations.ts

```diff
@@ -393,41 +407,53 @@ export default class Conversations {
       contact = SignedPublicKeyBundle.fromLegacyBundle(contact)
     }
 
-    for (const sealedInvite of await this.client.listInvitations()) {
-      const isSamePeer =
-        sealedInvite.v1.header.recipient.equals(contact) ||
-        sealedInvite.v1.header.sender.equals(contact)
-      if (!isSamePeer) {
-        continue
-      }
-      try {
-        // Need to decode invite even without a context to ensure decryption succeeds and invite is valid
-        const invite = await sealedInvite.v1.getInvitation(this.client.keys)
-        // If the contexts match, return early
-        if (isMatchingContext(context, invite.context)) {
-          return await ConversationV2.create(
-            this.client,
-            invite,
-            sealedInvite.v1.header
-          )
+    // Define a function for matching V2 conversations
+    const matcherFn = (convo: Conversation) =>
+      convo.peerAddress === peerAddress &&
+      isMatchingContext(context, convo.context ?? undefined)
+
+    let v2Convo: Conversation | undefined
+    // Perform all future operations while holding the mutex
+    await this.v2Cache.load(
+      async ({ latestSeen, existing }): Promise<Conversation[]> => {
+        // First check the cache without doing a network request
+        const existingMatch = existing.find(matcherFn)
+        if (existingMatch) {
+          v2Convo = existingMatch
+          return []
         }
-      } catch (e) {
-        console.warn('Error decoding invite', e)
-      }
-    }
 
-    // If no existing invite, send a new one
-    const invitation = InvitationV1.createRandom(context)
-    const sealedInvite = await this.sendInvitation(
-      contact,
-      invitation,
-      new Date()
-    )
-    return ConversationV2.create(
-      this.client,
-      invitation,
-      sealedInvite.v1.header
+        // Next try and load new items into the cache from the network
+        const newItems = await this.v2ConversationLoader(latestSeen)
+        const newItemMatch = newItems.find(matcherFn)
+        // If one of those matches, return it to update the cache
+        if (newItemMatch) {
+          v2Convo = newItemMatch
+          return newItems
+        }
+
+        // If all else fails, create a new invite
+        const invitation = InvitationV1.createRandom(context)
+        const sealedInvite = await this.sendInvitation(
+          contact as SignedPublicKeyBundle,
+          invitation,
+          new Date()
+        )
+
+        v2Convo = await ConversationV2.create(
+          this.client,
+          invitation,
+          sealedInvite.v1.header
+        )
+
+        return [v2Convo]
+      }
     )
+
+    if (!v2Convo) {
```

Comment:
This is just there to keep the type checker happy. Should be impossible, since an error inside the loader would bubble up and halt execution before it gets here.

Reply:
I'd add this as a comment, it's not obvious.